File: ports/python/needprotocol/parse.py
"""Parse need events with dual-read (envelope / bare / legacy top-level).

Mirrors github.com/diogoX451/archon-need-protocol ParseNeedEvent so other
languages can implement against the same golden fixtures under
testdata/golden/.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any


@dataclass
class ExternalNeed:
    type: str = ""
    correlation_id: str = ""
    payload: Any = field(default_factory=dict)
    created_at: str = ""


@dataclass
class NeedEvent:
    schema_version: int = 0
    net_id: str = ""
    workflow_id: str = ""
    need: ExternalNeed = field(default_factory=ExternalNeed)


@dataclass
class WebhookBody:
    payload: Any = None


def _as_need(obj: dict[str, Any]) -> ExternalNeed:
    return ExternalNeed(
        type=str(obj.get("type") or ""),
        correlation_id=str(obj.get("correlation_id") or ""),
        payload=obj.get("payload") if obj.get("payload") is not None else {},
        created_at=str(obj.get("created_at") or ""),
    )


def parse_need_event(data: bytes | str | dict[str, Any]) -> NeedEvent:
    """Dual-read ParseNeedEvent compatible with Go goldens."""
    if isinstance(data, (bytes, bytearray)):
        obj = json.loads(data.decode("utf-8"))
    elif isinstance(data, str):
        obj = json.loads(data)
    else:
        obj = data

    # Outer envelope: {schema_version, payload: {...NeedEvent...}}
    if isinstance(obj, dict) and "payload" in obj and "need" not in obj and "correlation_id" not in obj:
        inner = obj["payload"]
        if isinstance(inner, str):
            try:
                inner = json.loads(inner)
            except json.JSONDecodeError:
                pass
        if isinstance(inner, dict) and ("need" in inner or "correlation_id" in inner or "type" in inner):
            obj = inner

    event = NeedEvent(
        schema_version=int(obj.get("schema_version") or 0),
        net_id=str(obj.get("net_id") or ""),
        workflow_id=str(obj.get("workflow_id") or ""),
    )
    if "need" in obj and isinstance(obj["need"], dict):
        event.need = _as_need(obj["need"])
    elif obj.get("correlation_id"):
        # legacy top-level ExternalNeed
        event.need = _as_need(obj)
    return event

File: ports/typescript/src/parse.ts
/**
 * Dual-read ParseNeedEvent compatible with Go goldens under testdata/golden/.
 * Mirrors github.com/diogoX451/archon-need-protocol.
 */

export interface ExternalNeed {
  type: string;
  correlation_id: string;
  payload: unknown;
  created_at?: string;
}

export interface NeedEvent {
  schema_version: number;
  net_id: string;
  workflow_id: string;
  need: ExternalNeed;
}

export interface WebhookBody {
  payload: unknown;
}

function asNeed(obj: Record<string, unknown>): ExternalNeed {
  return {
    type: String(obj.type ?? ""),
    correlation_id: String(obj.correlation_id ?? ""),
    payload: obj.payload ?? {},
    created_at: obj.created_at ? String(obj.created_at) : undefined,
  };
}

export function parseNeedEvent(data: string | Uint8Array | object): NeedEvent {
  let obj: Record<string, unknown>;
  if (typeof data === "string") {
    obj = JSON.parse(data) as Record<string, unknown>;
  } else if (data instanceof Uint8Array) {
    obj = JSON.parse(new TextDecoder().decode(data)) as Record<string, unknown>;
  } else {
    obj = data as Record<string, unknown>;
  }

  // Outer envelope: { schema_version, payload: NeedEvent }
  if (
    obj &&
    typeof obj === "object" &&
    "payload" in obj &&
    !("need" in obj) &&
    !("correlation_id" in obj)
  ) {
    let inner = obj.payload;
    if (typeof inner === "string") {
      try {
        inner = JSON.parse(inner);
      } catch {
        /* keep */
      }
    }
    if (
      inner &&
      typeof inner === "object" &&
      ("need" in (inner as object) ||
        "correlation_id" in (inner as object) ||
        "type" in (inner as object))
    ) {
      obj = inner as Record<string, unknown>;
    }
  }

  const event: NeedEvent = {
    schema_version: Number(obj.schema_version ?? 0),
    net_id: String(obj.net_id ?? ""),
    workflow_id: String(obj.workflow_id ?? ""),
    need: { type: "", correlation_id: "", payload: {} },
  };

  if (obj.need && typeof obj.need === "object") {
    event.need = asNeed(obj.need as Record<string, unknown>);
  } else if (obj.correlation_id) {
    event.need = asNeed(obj);
  }
  return event;
}
